plutus-core/plutus-ir/src/PlutusIR/Purity.hs

```diff
@@ -84,6 +85,9 @@ isPure ver varStrictness = go
             TyAbs {} -> True
             Constant {} -> True
             IWrap _ _ _ t -> go t
+            Let _ _ bs t -> all isPureBinding bs && go t
+            Apply _ (LamAbs _ _ _ body) arg -> go body && go arg
+            TyInst _ (TyAbs _ _ _ body) _ -> go body
```

Comment:
I'm unsure about these cases. I'm tempted to say we should just leave it up to `Beta` to turn such things into let-bindings, and leave them alone here. Generally there's a question of how smart this function should be. Unsure though.

Reply:
In most of our test cases we either don't run simplifiers or run 4 iterations (which include both beta and inliner), so removing the `Apply` and `TyInst` cases doesn't affect anything (other than `single.golden`).

Reply:
So why include them?

Reply:
Removed.